### ISSUE TITLE
look into using geodist for distance calculations

### ISSUE THREAD

**Issue by njtierney** [njtierney/maxcovr#50]
https://cran.r-project.org/web/packages/geosphere/index.html

**Comment by mpadge** [comment]
Oh yeah, sorry, just wrote what i did on #51, and hadn't yet seen this one ...

**Comment by mpadge** [comment]
some motivation for you here:

``` r
devtools::load_all ("<...>/maxcovr", export_all = TRUE)
#> Loading maxcovr
suppressMessages (library (dplyr))
york_selected <- york %>% filter(grade == "I")
york_unselected <- york %>% filter(grade != "I")
distance_cutoff <- 100
user <- tibble::rowid_to_column(york_crime, var = "user_id")
user_not_covered <- find_users_not_covered(york_selected,
                                           user,
                                           distance_cutoff)
proposed_facility <- york_unselected

knitr::kable (rbenchmark::benchmark (
    A <- binary_distance_matrix(facility = proposed_facility,
                                user = user_not_covered,
                                distance_cutoff = distance_cutoff),
    d1 <- which (geodist::geodist (user_not_covered, proposed_facility, measure = "cheap") < distance_cutoff),
    d2 <- which (geodist::geodist (user_not_covered, proposed_facility, measure = "haversine") < distance_cutoff),
    d3 <- which (geodist::geodist (user_not_covered, proposed_facility, measure = "geodesic") < distance_cutoff),
replications = 2, order = "relative"))
```

<table>
<thead>
<tr class="header">
<th></th>
<th style="text-align: left;">test</th>
<th style="text-align: right;">replications</th>
<th style="text-align: right;">elapsed</th>
<th style="text-align: right;">relative</th>
<th style="text-align: right;">user.self</th>
<th style="text-align: right;">sys.self</th>
<th style="text-align: right;">user.child</th>
<th style="text-align: right;">sys.child</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>2</td>
<td style="text-align: left;">d1 &lt;- which(geodist::geodist(user_not_covered, proposed_facility, measure = “cheap”) &lt; distance_cutoff)</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0.313</td>
<td style="text-align: right;">1.000</td>
<td style="text-align: right;">0.252</td>
<td style="text-align: right;">0.059</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td>3</td>
<td style="text-align: left;">d2 &lt;- which(geodist::geodist(user_not_covered, proposed_facility, measure = “haversine”) &lt; distance_cutoff)</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0.440</td>
<td style="text-align: right;">1.406</td>
<td style="text-align: right;">0.379</td>
<td style="text-align: right;">0.060</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td>1</td>
<td style="text-align: left;">A &lt;- binary_distance_matrix(facility = proposed_facility, user = user_not_covered, distance_cutoff = distance_cutoff)</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0.644</td>
<td style="text-align: right;">2.058</td>
<td style="text-align: right;">0.584</td>
<td style="text-align: right;">0.060</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td>4</td>
<td style="text-align: left;">d3 &lt;- which(geodist::geodist(user_not_covered, proposed_facility, measure = “geodesic”) &lt; distance_cutoff)</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">6.716</td>
<td style="text-align: right;">21.457</td>
<td style="text-align: right;">6.657</td>
<td style="text-align: right;">0.053</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>

<sup>Created on 2019-06-05 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Your C++ code is not too bad, but geodist is still faster, and offers the distinct advantage of multiple measures, including nanometre-accuracy geodesic, which some folk may consider important. This would simply involve replacing your `binary_distance_matrix` function with the code used above of:
``` r
which (geodist::geodist (...) < distance_cutoff)
```
Shall i PR that too?

**Comment by mpadge** [comment]
Note also that your radius is not the "official" WSG84 standard as in [`geodist`](https://github.com/hypertidy/geodist/blob/master/src/WSG84-defs.h), and most other places, and so `geodist(..., measure = "haversine")` does not give exactly the same results as your code. But just switching to `geodist` will resolve that anyway ...

**Comment by njtierney** [comment]
Hi @mpadge !

That is indeed great motivation!

I would love a PR to implement this - I think that the unit tests should help make sure not too much else breaks, I think that one problem may be that the format of `geodist()` and `binary_distance_matrix()` , so maybe there might need to be some syntactic sugar around that?

**Comment by mpadge** [comment]
I'll gladly get on to it, but wait for [this PR](https://github.com/njtierney/maxcovr/pull/78) first, noting especially that travis is currently failing because neither `gurobi` nor `Rglpk` can be installed. Once you've fixed that, then I'll get moving. :+1:

**Comment by njtierney** [comment]
Yeah so I'm not sure what to do about `gurobi` - it's an internal package from gurobi that isn't on CRAN or GitHub. I'll make another issue for that.
